gcloud/bigquery/job.py

```diff
@@ -888,14 +924,49 @@ class QueryJob(_AsyncJob):
     :type client: :class:`gcloud.bigquery.client.Client`
     :param client: A client which holds credentials and project configuration
                    for the dataset (which requires a project).
+
+    :type udf_resources: tuple
+    :param udf_resources: An iterable of
+                        :class:`gcloud.bigquery.query.UDFResource`
+                        (empty by default)
     """
     _JOB_TYPE = 'query'
+    _UDF_KEY = 'userDefinedFunctionResources'
+    _udf_resources = None
 
-    def __init__(self, name, query, client):
+    def __init__(self, name, query, client, udf_resources=()):
         super(QueryJob, self).__init__(name, client)
         self.query = query
+        self.udf_resources = udf_resources
         self._configuration = _AsyncQueryConfiguration()
 
+    @property
+    def udf_resources(self):
+        """Property for list of UDF resources attached to a query
+        See
+        https://cloud.google.com/bigquery/user-defined-functions#api
+        """
+        if not self._udf_resources:
+            return []
+        if not all(isinstance(u, UDFResource) for u in self._udf_resources):
+            raise ValueError("udf items must be UDFResource")
+        return list(self._udf_resources)
+
+    @udf_resources.setter
+    def udf_resources(self, value):
+        """Update queries UDF resources
+
+        :type value: list of :class:`UDFResource`
+        :param value: an object which defines the type and value of a resource
+
+        :raises: TypeError if 'value' is not a sequence, or ValueError if
+                 any item in the sequence is not a UDFResource
+        """
+        if not all(isinstance(u, UDFResource) for u in value):
+            raise ValueError("udf items must be UDFResource")
+        if len(value) > 0:
+            self._udf_resources = tuple(value)
```

Comment:
The empty tuple seems like a good initial value, too:  it would simplify the getter even further to:

```python
        return list(self._udf_resources)
```

The test below for whether to add the UDF resources to the query/job resource would need to accommodate that, of course.